File: src/test/scala/com/ajjpj/amapper/core/PathTest.scala
package com.ajjpj.amapper.core

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.matchers.ShouldMatchers
import org.scalatest.FunSuite

/**
 * @author arno
 */
@RunWith(classOf[JUnitRunner])
class PathTest extends FunSuite with ShouldMatchers {
  test("PathBuilder") {
    val builder = new PathBuilder() + SimplePathSegment("a") + SimplePathSegment("b")
    builder.build should equal (APath(List(SimplePathSegment("a"), SimplePathSegment("b"))))

    val b1 = builder + SimplePathSegment("c1")
    val b2 = builder + SimplePathSegment("c2")
    b1.build should equal (APath(List(SimplePathSegment("a"), SimplePathSegment("b"), SimplePathSegment("c1"))))
    b2.build should equal (APath(List(SimplePathSegment("a"), SimplePathSegment("b"), SimplePathSegment("c2"))))
  }

  test("toString") {
    APath(List(SimplePathSegment("a"), ParameterizedPathSegment("b", "keyB"), IndexedPathSegment("c", 3))).toString should equal ("Path{a.b(keyB).c[3]}")
  }

  test("decompose") {
    val p = APath(List(SimplePathSegment("a"), ParameterizedPathSegment("b", "keyB"), IndexedPathSegment("c", 3)))

    p.last should equal(IndexedPathSegment("c", 3))
    p.parent should equal (APath(List(SimplePathSegment("a"), ParameterizedPathSegment("b", "keyB"))))

    val shortPath = APath(List(SimplePathSegment("a")))
    shortPath.parent should equal (APath(Nil))
    shortPath.last should equal (SimplePathSegment("a"))
  }
}

File: src/test/java/com/ajjpj/amapper/core/PathTest.java
package com.ajjpj.amapper.core;

import com.ajjpj.amapper.core.path.APath;
import com.ajjpj.amapper.core.path.APathSegment;
import org.junit.Test;

import static org.junit.Assert.*;


/**
 * @author arno
 */
public class PathTest {
    @Test
    public void testBuild() {
        final APath p = APath.EMPTY.withChild(APathSegment.simple("a")).withChild(APathSegment.simple("b"));
        assertEquals(APath.fromSegments(APathSegment.simple("a"), APathSegment.simple("b")), p);

        final APath p1 = p.withChild(APathSegment.simple("c1"));
        final APath p2 = p.withChild(APathSegment.simple("c2"));

        assertEquals(APath.fromSegments(APathSegment.simple("a"), APathSegment.simple("b"), APathSegment.simple("c1")), p1);
        assertEquals(APath.fromSegments(APathSegment.simple("a"), APathSegment.simple("b"), APathSegment.simple("c2")), p2);
    }

    @Test
    public void testToString() {
        assertEquals("APath{a.b[keyB]}", APath.fromSegments(APathSegment.simple("a"), APathSegment.parameterized("b", "keyB")).toString());
    }

    @Test
    public void testDecompose() {
        final APath p = APath.fromSegments(APathSegment.simple("a"), APathSegment.parameterized("b", "keyB"));

        assertEquals(APathSegment.parameterized("b", "keyB"), p.getLastSegment());
        assertEquals(APath.fromSegments(APathSegment.simple("a")), p.getParent());

        assertEquals(APathSegment.simple("a"), p.getParent().getLastSegment());

        assertEquals(APath.EMPTY, p.getParent().getParent());
    }
}
